### ISSUE TITLE
Multiline works unpredictably

### ISSUE THREAD

**Issue by bcarrell** [abo-abo/lispy#35]
Here's an example in Clojure:

`-|-(= 1 1)`

Multiline transforms this into:
```
(=
 1
 1)
```

But in this case...

`-|-(= {:a 1} {:a 1})`

Multiline transforms this into:

```
(= {:a 1}
   {:a 1})
```

This can be quite confusing on more complicated structures.  Vectors use the same behavior as the integer example above, whereas lists seem to follow the map structure in the second one.

**Comment by tuhdo** [comment]
@abo-abo The commands one-line/multi-line are pretty stable now, you can check it [here](https://github.com/tuhdo/semantic-refactor/blob/master/srefactor-elisp.el). It solves all the problems with @bcarrell had above and in my issue. However, since he uses Clojure, he may need to install [clojure-semantic](https://github.com/kototama/clojure-semantic) and I'm not sure how well the grammar in `clojure-semantic` is written. However, since it has its own grammar file (which Emacs Lisp does not have), I hope it would be better.

@bcarrell if you want to try it own, paste the above file and bind the two commands to something else. As said, if you want to use it with Clojure, give `clojure-semantic` a try. However, if you want to test with Emacs Lisp first, then remember to run `(semantic-default-elisp-setup)` before activating `(semantic-mode 1)`.

**Comment by tuhdo** [comment]
@bcarrell sorry to say but too bad, the support for `clojure-semanti` is still not enough. It cannot retrieve tag at current location (aside from function/variable definitions), so there's nothing `srefactor` can do at the moment.

**Comment by bcarrell** [comment]
@tuhdo No worries, thanks for trying!

**Comment by tuhdo** [comment]
Actually, there's still hope, It uses a different parser (the Wisent one), but exposes similar utility APIs. I get it to do basic oneline/multiline now, but not for complex expression yet.

**Comment by tuhdo** [comment]
@bcarrell Now Srefactor also supports Clojure with [srefactor-clojure](https://github.com/tuhdo/semantic-refactor/blob/master/srefactor-clojure.el). You can use the two commands:

- srefactor-clojure-one-line
- srefactor-clojure-multi-line

They will convert a sexp back and forth without breaking anything. And it solves all your use case above. However, before using the command, you need to install [clojure-semantic](https://github.com/kototama/clojure-semantic). Follow the instructions to install it, with one additional steps:

- Press `M-:`, then eval `(wisent-clojure-default-setup)`. Then disable then enable `semantic-mode` again for the changes to take effect, since CEDET that came with Emacs 24.4 has problem with Lisp mode activation in general. We have to run the setup function manually and re-enable `semantic-mode`.

After that, enjoy other Semantic stuffs outside of these two commands.

**Comment by tuhdo** [comment]
You can see the demo [here](https://github.com/tuhdo/semantic-refactor/blob/master/srefactor-demos/demos-elisp.org). It can work reliably on any sexp now, and can skip symbols before inserting newline i.e. you don't want any newline at the start of `defun`, only after the argument list, or `progn` usually has a newline right after it; srefactor-elisp-one-line/multi-line can distinguish between these cases of special symbols, as you can see in the demo.

**Comment by abo-abo** [comment]
> reliably

Doesn't really look reliable to me. It's surprising that `if` doesn't
get one-lined. And I have no idea why someone would want that.

**Comment by tuhdo** [comment]
> Doesn't really look reliable to me. It's surprising that if doesn't
get one-lined. And I have no idea why someone would want that.

Currently it only transform on *sexp* line (not text line). All the sexp of the same level are turned into one line, but not the inner one. To do that, you have to visit the inner sexp and run the command. I'm working on the recursive version, so you can format the whole sexp down to every child node, so you can perform something like `srefactor-format-defun`.

**Comment by abo-abo** [comment]
> I'm working on the recursive version

Now that one wouldn't be useless. You can copy my implementation, [lispy-multiline-1](https://github.com/abo-abo/lispy/blob/master/lispy.el#L2493), which doesn't need Semantic.

**Comment by abo-abo** [comment]
Please give <kbd>T</kbd> a test. It's based on the rigorous idea of adding a newline after each
thing, but modified with pattern matching for familiar constructs, like `defun`, `defmacro` etc.

Support can be added for Clojure, as well as a way to customize the patterns.
For that, raise another issue with the specific behavior that you would like.

**Comment by tuhdo** [comment]
> Now that one wouldn't be useless. You can copy my implementation, lispy-multiline-1, which doesn't need Semantic.

I tested it on a fairly large defun. It got messed up badly. What I meant is formatting, not just turning one-line/multi-line. It means, it must format indent properly and account for `fill-column`.

**Comment by abo-abo** [comment]
> I tested it on a fairly large defun. It got messed up badly.

Please report a bug in that case.

**Comment by tuhdo** [comment]
I expect it should be able to do something like [the demo here](https://github.com/tuhdo/semantic-refactor/blob/master/srefactor-demos/demos-elisp.org). That is, you can format the whole buffer reliable, transform back and forth between online and multiline without deforming between each successive call.

**Comment by tuhdo** [comment]
@bcarrell now you do not need `clojure-semantic` to use my formatting tool anymore. A demo: 
[Formatting whole buffer demo](http://i.imgur.com/BNWSL7W.gifv) in Emacs Lisp (applicable in Common Lisp and Scheme and Clojure as well). All you need to do uses these commands:

- srefactor-lisp-format-buffer: format whole buffer
- srefactor-lisp-format-defun: format current defun cursor is in
- srefactor-lisp-one-line: turn the current sexp of the same level into one line; with prefix argument, recursively turn all inner sexps into one line.
- srefactor-lisp-multi-line: format the current sexp cursor is in.

No additional setup is needed.
